### ISSUE TITLE
Blocks rescanned at the PD startup

### ISSUE THREAD

**Issue by bacoinin** [particl/particl-desktop#1778]
**Describe the bug**
After restarting the PD v3 client, the application boots properly the Particl Core and the ZMQ service (both confirmed via ss -ntlp) and then stays for 10-20 min at the splash screen that says "Starting: Application ready". After the time elapses the application is launched. 

The particl/testnet/debug.log shows the following:
```
2021-05-25T21:08:43Z Using BerkeleyDB version Berkeley DB 4.8.30: (April  9, 2010)
2021-05-25T21:08:43Z Using wallet /home/mba/.particl/testnet/profiles/DEFAULT/particl-market
2021-05-25T21:08:43Z BerkeleyEnvironment::Open: LogDir=/home/mba/.particl/testnet/profiles/DEFAULT/particl-market/database ErrorFile=/home/mba/.particl/testnet/profiles/DEFAULT/particl-market/db.log
2021-05-25T21:08:43Z init message: Loading wallet...
2021-05-25T21:08:43Z BerkeleyEnvironment::Open: LogDir=/home/mba/.particl/testnet/profiles/DEFAULT/particl-market/database ErrorFile=/home/mba/.particl/testnet/profiles/DEFAULT/particl-market/db.log
2021-05-25T21:08:43Z [profiles/DEFAULT/particl-market] Loading master ext key.
2021-05-25T21:08:43Z [profiles/DEFAULT/particl-market] Loading ext accounts.
2021-05-25T21:08:43Z [profiles/DEFAULT/particl-market] Loading ext account packs.
2021-05-25T21:08:43Z [profiles/DEFAULT/particl-market] Preparing Lookahead pools.
2021-05-25T21:08:43Z [profiles/DEFAULT/particl-market] mapRecords.size() = 347
2021-05-25T21:08:43Z Warning: Error reading /home/mba/.particl/testnet/profiles/DEFAULT/particl-market/wallet.dat! All keys read correctly, but transaction data or address book entries might be missing or incorrect.
2021-05-25T21:08:43Z [profiles/DEFAULT/particl-market] Wallet completed loading in              66ms
2021-05-25T21:08:43Z init message: Rescanning...
2021-05-25T21:08:43Z [profiles/DEFAULT/particl-market] Rescanning last 891506 blocks (from block 0)...
2021-05-25T21:08:43Z [profiles/DEFAULT/particl-market] Rescan started from block 0000594ada5310b367443ee0afd4fa3d0bbd5850ea4e33cdc7d6a904a7ec7c90...
2021-05-25T21:08:46Z [profiles/DEFAULT/particl-market] ScanForOwnedOutputs: txn cd2aa9768fad5370ef8e2304e582203c7422665c5974ec1707cbf5ff3e026cae, malformed data output 0.
2021-05-25T21:09:43Z [profiles/DEFAULT/particl-market] Still rescanning. At block 85511. Progress=0.093174
2021-05-25T21:10:43Z [profiles/DEFAULT/particl-market] Still rescanning. At block 113879. Progress=0.127697
2021-05-25T21:11:45Z [profiles/DEFAULT/particl-market] Still rescanning. At block 113882. Progress=0.127930
2021-05-25T21:13:10Z [profiles/DEFAULT/particl-market] Still rescanning. At block 113890. Progress=0.128469
2021-05-25T21:14:10Z [profiles/DEFAULT/particl-market] Still rescanning. At block 113894. Progress=0.128867
2021-05-25T21:15:10Z [profiles/DEFAULT/particl-market] Still rescanning. At block 113903. Progress=0.129213
2021-05-25T21:16:10Z [profiles/DEFAULT/particl-market] Still rescanning. At block 114122. Progress=0.129831
2021-05-25T21:17:10Z [profiles/DEFAULT/particl-market] Still rescanning. At block 175296. Progress=0.204766
2021-05-25T21:18:10Z [profiles/DEFAULT/particl-market] Still rescanning. At block 337133. Progress=0.389427
2021-05-25T21:19:10Z [profiles/DEFAULT/particl-market] Still rescanning. At block 395699. Progress=0.459116
2021-05-25T21:20:10Z [profiles/DEFAULT/particl-market] Still rescanning. At block 419643. Progress=0.487594
```

**To Reproduce**
Not sure. You could try to force kill the particld and  then restart the PD v3

**Expected behavior**
Normal start-up without the need to re-scan the blockchain

**Versions (please complete the following information):**
 - OS: Linux Debian 10
 - Particl Desktop v3.0 RC-test003
 - Particl Core: 0.19.2.9

**Additional context**
I have encountered the same problem in the past on the RC-test002, both on Windows and Linux. If I am not mistaken @coybigpart also had the same issue

**Comment by zaSmilingIdiot** [comment]
When the splash screen displays the "Starting: Application ready" status, Particl Desktop (PD) has at this point confirmed that particl-core has started and for all relevant purposes PD should be able to connect to a running particl-core instance. 
However, at this moment, PD starts polling that particl-core instance to (a) determine that we can in fact, actually connect successfully, and (b) to fetch relevant initial network information (such as whether the core instance is serving up the test or main chain), which is necessary to some functionality prior to actually presenting the interactive nature of PD itself.

It should be noted that the output shown in the snippet is all particl-core output, of which Particl Desktop has no control over :) ~Why this is important: at the time that core is performing this scan, its likely (to be confirmed) that it is not in fact responding to such information gathering requests, ie: particl-core's rpc service may not be responding to the relevant rpc requests or possibly returning an error response indicating that it is not ready yet. Without establishing a valid connection and retrieving the relevant start up information, PD cannot present the rest of the interactive interface to the user - nothing would effectively work.~

~The difficulty here is obtaining the details about whether the rpc service is available or not (and whether its still starting up or otherwise not ready yet), as well as determining exactly what core is doing while the rpc service is unavailable (in order to provide feedback to the user). There is a longer term plan to overhaul the connectivity mechanism to particl-core and how feedback is presented to the user based on changes implemented in v3.0, which would likely address this type of issue.~

~In the meanwhile, a temporary solution may be to simply update the relevant status message to first indicate that the connection to core is still pending, rather than indicating that the application is ready. And only once a rpc response is successfully received from core, then the "application ready" status is presented... although once a successful rpc response is received from core then PD is immediately useful, so the latter "Application ready" status may be irrelevant.~

I should have confirmed the above position before commenting: my previous commentary is not exactly the current case with v3.0. The 1st rpc response that PD receives back successfully from particl-core that contains the necessary details is in fact what results in PD actually displaying the "Application ready" status.

Looking into it in a little more detail this time around, it would appear that in this sampled case, the even that progresses the UI to allow for user interactivity either is not firing correctly, or the event is not being detected in the correct location correctly. The former may be more likely, but I'll continue looking into this further.

**Comment by zaSmilingIdiot** [comment]
Just noting (mostly for myself) that in the test cases attempting to reproduce this, each starting with a new core datadir such that the blockchain requires sync and download, the UI appeared to respond correctly whether before, during or after the blockheaders have synced. While I have not been able to reproduce this so far, it would suggest that the actual particl-core event(s) of syncing blockchain headers are not a direct factor of the UI not updating.

**Comment by zaSmilingIdiot** [comment]
A potential fix has been added for this, although it will need to be checked and tested to ensure that it does in fact, solve the original reported issue.

**Comment by bacoinin** [comment]
I did a bit more testing... 

I always run the PD via a terminal, both on Windows and Linux. Often when I close the PD via the GUI or ctrl+c on the terminal and the particld continues running in the background. I love that because when I start again the PD, it starts practically instantly. (Note: maybe this could be a feature? I would love that)

The above issue that I logged occurred last time when I wanted the PD to update to the newest core 0.19.2.10 so I had to forcefully kill the deamon (kill "pid").  So after restarting the GUI it downloaded the new core but when booting it, for some reason the core printed "All keys read correctly, but transaction data or address book entries might be missing or incorrect." and decided that a blockchain rescan was needed...thus the 20+min delaying in the application start up.

Could it be linked to the PD failing to shutdown the particld properly and me forcefully killing it manually after some time has passed ? Maybe @tecnovert has the answer

**Comment by zaSmilingIdiot** [comment]
Terminating particl-core that way would possibly lead to it determining that a re-index or re-sync is necessary. 

However, this is independent of what is happening in the PD GUI: regardless of whether the block headers are being re-synced or not, nor whether the blocks are being redownloaded/re-synced/re-indexed or not, all PD is waiting for is the particl-core's rpc service to become available. Initially I made the assumption that the "hanging" of PD at the "Application ready" status was because particl-core's rpc service was not yet available. However that was incorrect and looking through the PD code again confirmed that the "Application ready" status in PD is only displayed after confirmation is received in PD that the rpc service is available and the relevant applicable blockchain information has been obtained via rpc request. At that point, there is no further core interaction required in order to proceed to displaying the Wallet and Market modules. In fact, the event emitted (event A) to indicate that a successful rpc response has been received from core, performs all of the following concurrently:
- dispatch an event (event B) to update the loading status to "Application ready"
- dispatch an event (event C) to load effect the loading of the UI interaction

To reiterate: both of those events B and C emitted are specific to PD itself, and are emitted concurrently.
Since its known that PD displays the "Application ready" status, which is only caused as a result of event B, and B is only emitted when A is successful, then A must have emitted. The problem here is that either event C was not emitted correctly, or the listener waiting for it did not respond correctly or did not receive the event listener correctly. To this end, I've implemented a change in the emission of event C as I believe there may be an edge case where its possible that it might not have been emitted correctly. Hence the need to test whether this was the case.

Side note: as part of an attempt to reproduce the error I did in fact, terminate core suddenly, which resulted in a similar redownload of the blockchain on restart of core. However, in my test case, PD functioned correctly and I couldn't re-produce the error this way. Which is why I haven't marked this issue as resolved yet, and feel some additional testing using the potential fix would be helpful.

Also, for reference, its not typically a good idea to terminate PD this way :) There is some cleanup that PD goes through on shutdown. If its been working for you then great :D I just wouldn't recommend most people do this. 

If the goal is to have particl-core running in the background the whole time and then just letting PD connect to it when necessary, that is possible to do this by
- configuring particl-core to run with `daemon=1` (or pass the `-daemon` flag to the particld binary on execution),
- start core by running the `particld` executable (don't forget to add the `-testnet` flag if running on the test chain).
When PD starts the next time it'll connect to the running daemon as expected and experienced already. But when closing PD (correctly) it will not try to terminate the core daemon because PD was not the service that started it.
Thats the preferred way of achieving the goal of connecting PD to an already running core instance.

**Comment by tecnovert** [comment]
> Could it be linked to the PD failing to shutdown the particld properly and me forcefully killing it manually after some time has passed ? Maybe @tecnovert has the answer

Yes, force killing the daemon will usually cause a rescan on the next startup, as the databases won't have shutdown correctly and may not have been synced to disk.

You can get the daemon to shutdown cleanly by passing SIGINT instead of SIGKILL, eg:
`pkill -2 particld``

**Comment by zaSmilingIdiot** [comment]
Closing: fixed for mainnet release of v3.0.0
